libsolidity/analysis/TypeChecker.cpp

```diff
@@ -3820,6 +3919,148 @@ void TypeChecker::endVisit(UsingForDirective const& _usingFor)
 					": " +  result.message()
 				)
 			);
+		else if (operator_)
+		{
+			if (!_usingFor.typeName()->annotation().type->typeDefinition())
+			{
+				m_errorReporter.typeError(
+					5332_error,
+					path->location(),
+					"Operators can only be implemented for user-defined types and not for contracts."
+				);
+				continue;
```

Comment:
But isn't that because the error messages are telling you what do instead of telling you what's wrong? There's usually more than one way to fix an error and if the message looks suboptimal here, I'm sure we'll find other situations where the suggestion is wrong and confusing.

```solidity
// TypeError 5332: (7-8): Operators can only be implemented for user-defined value types and structs.
// TypeError xxxx: (7-8): Operator parameter type mismatch. The types of parameters of function "f" bound to operator + on type "contract C" do not match that type.
// TypeError yyyy: (7-8): Operator return type mismatch. The type of return value of function "f" bound to operator + on type "contract C" does not match that type.
```

That's basically what I'm getting at in https://github.com/ethereum/solidity/pull/12362#discussion_r976879326.

Reply:
I think that's more useful. If the second type was meant to be `Uint` instead, user might want to fix it in both places.